### PR TITLE
vlang: init at 0.1.16

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
V is an interesting new language, might as well have a package for it.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/v/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    installPhase = ''
+        mkdir -p $out/{bin,share}
+        ./vc -o $out/bin/v .
+        cp -r ../examples $out/share
+    '';
+
+    meta = with stdenv.lib; {
+        homepage = "https://vlang.io/";
+        description = "Simple, fast, safe, compiled language for developing maintainable software ";
```
Comment (JohnAZoidberg) [inline pkgs/development/compilers/v/default.nix]:
```suggestion
        description = "Simple, fast, safe, compiled language for developing maintainable software";
```

File: pkgs/development/compilers/v/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        rev = "v${version}";
+        sha256 = "0jdzlgmrkvwwnwsnd8m5zsqyagxiaval45q1mqqdrxhwsg54b8b3";
+    };
+
+    buildInputs = [ glfw freetype curl ];
+
+    buildPhase = ''
+        export HOME=`pwd`
+        mkdir -p code/v
+        cp -r `ls -A | grep -v "code"` code/v
```
Comment (JohnAZoidberg) [inline pkgs/development/compilers/v/default.nix]: Do you need to create a new directory? Their README recommends to build in the `compiler` subdirectory. You could use `sourceRoot` to go there.
Especially this line looks  quite complicated at first glance (If you use `mv` you don't need `grep -v`, I think)

**Review by JohnAZoidberg — COMMENTED** [review]
I tried it and could successfully run the tetris example in a nix-shell with `freetype curl glfw`. Because of vlang/v#542 I also had to copy the font to my working directory.

**Comment by JohnAZoidberg** [comment]
If it takes them a long time to figure out vlang/v#457, maybe we could solve it with a wrapper, at least for the standard library.

File: pkgs/development/compilers/v/default.nix (hunk 1 above)
Comment (Chiiruno) [reply pkgs/development/compilers/v/default.nix]: Oh, that's embarrassing.

File: pkgs/development/compilers/v/default.nix (hunk 2 above)
Comment (Chiiruno) [reply pkgs/development/compilers/v/default.nix]: Yes, home must be set, and there must be a code directory, in which the V directory goes that contains the files and sub-directories.
See: https://github.com/vlang/v/issues/457
That `cp` command copies everything but the `code` directory, is there a better command for this?

**Comment by Chiiruno** [comment]
This changed quick.
https://github.com/vlang/v/commit/5a469c2e376480248493329309c71c2a19e56e88
Gonna see what I can do.

**Comment by Chiiruno** [comment]
Waiting until the releases page and v.c is updated, don't want to flip flop between revision and release revision if possible.

File: pkgs/development/compilers/v/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
+{ stdenv, fetchurl, fetchFromGitHub, glfw, freetype, curl }:
+
+let
+    # V compiler's source translated to C for bootstrapping.
+    vc = fetchurl {
+        url = "https://github.com/vlang/v/releases/download/v0.1.0/v.c";
```
Comment (jtojnar) [inline pkgs/development/compilers/v/default.nix]: This does not seem to exist.

File: pkgs/development/compilers/v/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        sha256 = "0vm5f0538c4gflj3k0kpin2f3iw7r4xi75cmiincz1913d1hbkh7";
+    };
+in stdenv.mkDerivation rec {
+    name = "v-${version}";
```
Comment (jtojnar) [inline pkgs/development/compilers/v/default.nix]: You can just use 

```suggestion
    pname = "v";
```

File: pkgs/development/compilers/v/default.nix (hunk 3 above)
Comment (Chiiruno) [reply pkgs/development/compilers/v/default.nix]: Yes, seems like something is happening upstream.
Will change when upstream changes.

File: pkgs/development/compilers/v/default.nix (hunk 4 above)
Comment (Chiiruno) [reply pkgs/development/compilers/v/default.nix]: So that's what that's about.
Thanks.

**Comment by vcunat** [comment]
A bit meta-question: isn't "v" a bit too short/generic identifier?  In particular, it seems that [a couple distros](https://repology.org/project/v) use it already for a different project: https://github.com/rupa/v

**Comment by Chiiruno** [comment]
Since V is a language, changing it to vlang should be fine IMO.

**Comment by vcunat** [comment]
:+1: That's also supported by them using https://github.com/vlang.

File: pkgs/development/compilers/v/default.nix (hunk 2 above)
Comment (Chiiruno) [reply pkgs/development/compilers/v/default.nix]: No need for code directory anymore.

**Comment by Chiiruno** [comment]
Get this problem with the bootstrapped V compiler, tested locally and it doesn't happen, will make issue both here and on the V repo, not sure which is at fault.
```
Looks like you are running V for the first time.
Setting VROOT to "/build/source".
time__ticks
0
pass=2 fn=`time__ticks`
panic: time_lin.v:9
undefined: `double`
```

**Comment by Chiiruno** [comment]
Not a nix issue, 0.1.4 should be the version that _finally_ gets merged.

**Comment by Chiiruno** [comment]
Okay, things should be stable enough in this aspect for merge now.

**Comment by Chiiruno** [comment]
Oh whoops, forgot to change it to vlang instead of v, one sec.

**Comment by Chiiruno** [comment]
Double checked; Ready.

File: pkgs/development/compilers/vlang/default.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        rev = "v${version}";
+        sha256 = "0sip76rxpx5w7vy9s9vpib0ibfr5r2msfl81wcd966hg7p6cd5i6";
+    };
+
+    buildInputs = [ glfw freetype curl ];
+
+    buildPhase = ''
+        export HOME=`pwd`
+        cd compiler
+        cc -std=gnu11 -w -o v ${vc}/v.c
```
Comment (veprbl) [inline pkgs/development/compilers/vlang/default.nix]: This is not easily overridable. Consider replacing `${vc}` with `$vc` and moving `vc` to derivation attribute.

**Comment by Chiiruno** [comment]
Like this?

File: pkgs/development/compilers/vlang/default.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    buildPhase = ''
+        export HOME=`pwd`
+        cd compiler
+        cc -std=gnu11 -w -o v $vc/v.c
+    '';
+
+    installPhase = ''
+        mkdir -p $out/{bin,share}
+        ./v -o $out/bin/v .
```
Comment (veprbl) [inline pkgs/development/compilers/vlang/default.nix]: Technically, this is a part of the build phase.

File: pkgs/development/compilers/vlang/default.nix [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    vc = fetchFromGitHub {
+        owner = "vlang";
+        repo = "vc";
+        rev = "a0b6d1ac1c43380134c4c4f584b5abde8b438d1e";
+        sha256 = "0i5g38lfqbp8mzn8xas3sghmd6rdb2s8kr9ih25b0wxcjydmfv6h";
+    };
+
+    buildInputs = [ glfw freetype curl ];
+
+    buildPhase = ''
```
Comment (veprbl) [inline pkgs/development/compilers/vlang/default.nix]: This could use `runHook preBuild` and `runHook postBuild`. Same goes for the `installPhase`.

**Review by veprbl — COMMENTED** [review]
We usually indent our files with 2 spaces everywhere. This uses 3.

**Comment by Chiiruno** [comment]
Not sure how my tabs got messed up, but it's fixed now.

**Comment by Chiiruno** [comment]
@srhb The "wait for upstream" tag is no longer needed, could you remove it please?

**Comment by Chiiruno** [comment]
Executable says 0.1.9 when executed, but the release is 0.1.10, maybe a release issue.

**Comment by chreekat** [comment]
@Chiiruno , Sounds like another wait-for-upstream issue then?

**Comment by JohnAZoidberg** [comment]
Upstream is really fast and they're already at v0.1.12.

**Comment by Chiiruno** [comment]
Looks fine to me now.

**Comment by veprbl** [comment]
@GrahamcOfBorg eval

**Comment by veprbl** [comment]
@GrahamcOfBorg build vlang

**Comment by veprbl** [comment]
@Chiiruno
I could not make this work:
```
[nix-shell:~/.cache/nix-review/pr-63738]$ cp /nix/store/xz93v8lxlsjqak6r17c8qh7p2cbp07gx-vlang-0.1.12/share/examples/hello_world.v .

[nix-shell:~/.cache/nix-review/pr-63738]$ v hello_world.v 
Looks like you are running V for the first time.
V repo not found. Go to https://vlang.io to download V.zip 
or install V from source.
```

**Comment by Chiiruno** [comment]
To fix that, either we would need to trick the V compilation process into thinking it was compiled at the nix store directory, or have a way to set the V repo directory to the nix store directory.

**Comment by Chiiruno** [comment]
This may be fixed on the next update, with vlib being moved into /usr/lib.

**Comment by Chiiruno** [comment]
Should be ready.

**Comment by veprbl** [comment]
@GrahamcOfBorg build vlang

**Review by veprbl — APPROVED** [review]
Seems to work at least on darwin

**Pull request merged**